src/main/java/frc/robot/subsystems/pivot/Pivot.kt

```diff
@@ -1,181 +1,113 @@
+            .andThen(WaitUntilCommand { inputs.isAtHardstop })
+            .andThen(runOnce { io.setKnownPosition(Units.Degrees.of(90.0)); io.stop() })
 
-    /**
-     * Set a voltage amount to send to the pivot
-     * @param voltage The voltage to send to the pivot
-     */
-    fun setVoltage(voltage: Double) {
-        pivotMotor.setVoltage(voltage)
-    }
+    fun homeSensorless(threshold: Double = 5.5) =
+        runOnce { io.setRawVoltage(Units.Volts.of(-3.0)) }
+            .andThen(WaitUntilCommand { currentSpikeTracker.lastValue() > threshold })
+            .andThen(runOnce { io.setKnownPosition(Units.Degrees.of(90.0)); io.stop() })
 
-    /** Reset the absolute encoder */
-    fun resetEncoder() {
-        absoluteEncoder.reset()
-    }
+    fun sendToPosition(position: Measure<Angle>) = run { io.setTargetPosition(position) }
 
-    /**
-     * Set the position of the pivot using PID
-     * @param position The position to set the pivot to
-     */
-    fun setPIDPosition(position: Double) {
-        pivotPID.setReference(position, CANSparkBase.ControlType.kPosition)
-        println("Setting position to $position")
-    }
+    override fun periodic() {
+        io.updateInputs(inputs)
+        Logger.processInputs("pivot", inputs)
+
+        arm.angle = inputs.relativePosition into Units.Degrees
+
+//        io.setRawVoltage(Units.Volts.of(0.0), false)
 
-    /**
-     * Set the position of the pivot using PID with feedforward specialized to hold it in place
-     * @param position The position to set the pivot to
-     */
-    fun holdArm(position: Double){
-        pivotPID.setReference(position,
-            CANSparkBase.ControlType.kPosition,
-            0,
-            feedforward.calculate(Units.degreesToRadians(position), 0.0)
+        Logger.recordOutput(
+            "pivot/mechanismPose",
+            Pose3d.struct,
+            Pose3d(
+                originToPivot,
+                Rotation3d(0.0, -(inputs.relativePosition into Units.Radians), 0.0)
+            )
         )
-    }
 
-    /** Stop the pivot */
-    fun stop() {
-        pivotMotor.stopMotor()
+        Logger.recordOutput("pivot/mechanism2d", mechanism)
+
+        Logger.recordOutput("pivot/currentSpike", currentSpikeTracker.calculate(inputs.appliedCurrent into Units.Amps))
     }
 
-    fun manualControl(speed: Supplier<Double>) = run { setVoltage(speed.get()) }
-    fun home() = run { setRawSpeed(-0.2) }.until(::getHomingSensor)
-    fun setKnownPosition(position: Double) = runOnce { pivotMotor.encoder.position = position }
+    companion object {
+        const val GEARBOX_RATIO: Double = 36.0 / 1.0
+        const val PULLEY_RATIO: Double = 40.0 / 18.0
 
+        const val ABSOLUTE_ENCODER_PORT = 1
+        const val PIVOT_MOTOR_PORT = 16
+        const val HOMING_SENSOR_PORT = 3
+
+        const val INVERT = false
+
+        val originToPivot = Translation3d(
+            0.0,
+            0.0,
+            0.516847 + 0.042919
+        )
+    }
```

Comment:
I actually do prefer this convention for constants over the `Constants` object, but it does not match with the rest of the codebase or most wpi projects.

I would sign the petition to change that in a heartbeat though.

Reply:
It's something a few other kotlin teams do, our constants file got insanely crowded this year, I think the companion object is a good place to do it